File: DataStructures/BST.c
/******************************************************************************

Binary Tree
       4
      / \
     1   6
    /\
   5  2
  
  
  Binary Search Tree 
       8
      / \
     3   10
    /\
   1  6
   
*******************************************************************************/

#include <stdio.h>
#include <malloc.h>

struct node
{
  int data;
  struct node *left;
  struct node *right;
};

struct node *
createNode (int data)
{
  struct node *n;
  n = (struct node *) malloc (sizeof (struct node));
  n->data = data;
  n->left = NULL;
  n->right = NULL;
  return n;
}

void
preOrder (struct node *root)
{
  if (root != NULL)
    {
      printf ("%d", root->data);
      preOrder (root->left);
      preOrder (root->right);
    }
}

void
postOrder (struct node *root)
{
  if (root != NULL)
    {
      postOrder (root->left);
      postOrder (root->right);
      printf ("%d", root->data);
    }
}

void
inOrder (struct node *root)
{
  if (root != NULL)
    {
      inOrder (root->left);
      printf ("%d", root->data);
      inOrder (root->right);
    }
}

struct node*
searchBST (struct node *root, int key)
{
  if (root == NULL)
  {
      return NULL;
  }    
  if (key==root->data)
  {
  	  return root;
  }
  else if (key<root->data)
  {
  	  return searchBST (root->left, key);
  }
  else
  {
  	  return searchBST (root->right, key);
  }
}

struct node* searchBSTiter (struct node* root, int key)
{
  while (root!=NULL)
  {
    if(key==root->data)
    {
      return root;
    }
    else if(key < root->data) 
    {
      root = root->left;
    }
    else 
    {
      root = root->right;
    }
  }
  return NULL;
}

int
main ()
{
  // Creating Nodes for Binary Tree
  struct node *root = createNode (4);
  struct node *n1 = createNode (1);
  struct node *n2 = createNode (6);
  struct node *n3 = createNode (5);
  struct node *n4 = createNode (2);

  // Linking Nodes
  root->left = n1;
  root->right = n2;
  n1->left = n3;
  n1->right = n4;

  // Traversing Nodes
  preOrder (root);
  printf ("\n");
  postOrder (root);
  printf ("\n");
  inOrder (root);
  printf ("\n");

  // Creating Nodes for Binary Search Tree
  struct node *root2 = createNode (8);
  struct node *n12 = createNode (3);
  struct node *n22 = createNode (10);
  struct node *n32 = createNode (1);
  struct node *n42 = createNode (6);

  // Linking Nodes
  root2->left = n12;
  root2->right = n22;
  n12->left = n32;
  n12->right = n42;

  // Traversing Nodes
  preOrder (root2);
  printf ("\n");
  postOrder (root2);
  printf ("\n");
  inOrder (root2);
  printf ("\n");
  
  
  // Searching searchBST
  struct node* search = searchBST(root2, 4);
  if (search != NULL) 
  {
    printf("Element found : %d", search->data);
  }
  else
  {
    printf("Element not found");
  }
  
  
  printf("\n");
  
  
  // Searching BST Iteratively
  struct node* search2 = searchBSTiter(root2, 6);
  if (search2 != NULL) 
  {
    printf("Element found : %d", search2->data);
  }
  else
  {
    printf("Element not found");
  }

  return 0;
}

File: DataStructures/DS_in_JS/bst.js
/*
    A binary tree is a tree data structure in which each node has at at most two children. They are referred to as right child and left child.

    Binary Search Tree - value of left subtree must be smaller than root and value of right subtree is greater that root.
    In BST each node has at most two children.

    example of BST: 
           10
          /  \
        5     15
       / \     
      3   7

    BST Operations-

    Insertion
    Search
    DFS & BSF (Traversing)
    Deletion
    Contains

    Binary Search Tree Applications
    - Searching
    - Sorting
    - To implement abstract data types such as look up tables and priority queues.

    Linear data structures having one way to traverse but in hierarchal data structure have different ways to traversing.
    Traversing means visiting every node in the tree.

    Two methods are :-
    - Depth First Search
    - Breadth First Search

    DFS algorithm starts at the root node and explores as far as possible along each branch before backtracking.
    Visit the root node, visit all nodes in left subtree and visit all the nodes in the right subtree.
    Depending on the order in which we do this, there can be three types of DFS traversals :-  
    - pre-order (Root, Left, Right)
    - in-order (Left, Root, Right)
    - post-order (Left, Right, Root)

    In BFS, explore all nodes at the present depth prior to moving on to the nodes at the next depth level.
    Steps :- 
    - Create Queue
    - Enqueue the root node
    - As long as a node exists in the queue
        a. Dequeue the node from the front
        b. Read the node's value
        c. Enqueue the node's left child if it exists
        d. Enqueue the node's right child if it exists
*/

class Node {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class BST {
    constructor() {
        this.root = null;
    }

    isEmpty() {
        return this.root === null;
    }

    insert(value) {
        const newNode = new Node(value);
        if (this.isEmpty()) {
            this.root = newNode;
        } else {
            this.insertNode(this.root, newNode);
        }
    }

    insertNode(rootNode, newNode) {
        if (rootNode.value > newNode.value) {
            if (rootNode.left === null) {
                rootNode.left = newNode;
            } else {
                this.insertNode(rootNode.left, newNode);
            }
        } else {
            if (rootNode.right === null) {
                rootNode.right = newNode;
            } else {
                this.insertNode(rootNode.right, newNode);
            }
        }
    }

    find(value) {
        if (this.isEmpty()) {
            return null;
        } else {
            return this.findNode(this.root, value);
        }
    }

    findNode(rootNode, value) {
        if (rootNode.value === value) {
            return true;
        }
        if (rootNode.value > value) {
            if (rootNode.left === null) {
                return false;
            } else {
                return this.findNode(rootNode.left, value);
            }
        } else {
            if (rootNode.left === null) {
                return false;
            } else {
                return this.findNode(rootNode.right, value);
            }
        }
    }

    preOrder(root) {
        if (root) {
            console.log(root.value);
            this.preOrder(root.left);
            this.preOrder(root.right);
        }
    }
    inOrder(root) {
        if (root) {
            this.inOrder(root.left);
            console.log(root.value);
            this.inOrder(root.right);
        }
    }
    postOrder(root) {
        if (root) {
            this.postOrder(root.left);
            this.postOrder(root.right);
            console.log(root.value);
        }
    }
}

const bst = new BST();

console.log(bst.isEmpty());
bst.insert(10);
bst.insert(5);
bst.insert(15);
bst.insert(3);
bst.insert(7);
console.log(bst.isEmpty());

// bst.preOrder(bst.root);
// bst.inOrder(bst.root);
bst.postOrder(bst.root);